/*
* @license Apache-2.0
*
* Copyright (c) 2021 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import mapArguments = require( './index' );

/**
* Callback function.
*
* @param v - argument
* @returns result
*/
function clbk<T>( v: T ): T {
	return v;
}


// TESTS //

// The function returns a function...
{
	mapArguments<number>( ( x, y, z ): Array<number> => [ x, y, z ], clbk ); // $ExpectType (...args: number[]) => number[]
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], clbk, {} ); // $ExpectType (...args: string[]) => string[]
}

// The compiler throws an error if the function is provided a first argument other than a function...
{
	mapArguments( true, clbk ); // $ExpectError
	mapArguments( false, clbk ); // $ExpectError
	mapArguments( 5, clbk ); // $ExpectError
	mapArguments( [], clbk ); // $ExpectError
	mapArguments( {}, clbk ); // $ExpectError
	mapArguments( 'abc', clbk ); // $ExpectError

	mapArguments( true, clbk, {} ); // $ExpectError
	mapArguments( false, clbk, {} ); // $ExpectError
	mapArguments( 5, clbk, {} ); // $ExpectError
	mapArguments( [], clbk, {} ); // $ExpectError
	mapArguments( {}, clbk, {} ); // $ExpectError
	mapArguments( 'abc', clbk, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument other than a function...
{
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], '5' ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], true ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], false ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], 123 ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], null ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], [] ); // $ExpectError

	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], '5', {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], true, {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], false, {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], 123, {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], null, {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], {}, {} ); // $ExpectError
	mapArguments<string>( ( x, y, z ): Array<string> => [ x, y, z ], [], {} ); // $ExpectError
}

// The compiler throws an error if the function is provided more than three arguments...
{
	mapArguments<number>( ( x, y ): number => x + y, clbk, {}, 4 ); // $ExpectError
}
